File: campus-navigate-app/expo-app/src/components/map/actor/actorDirectionModel.ts
export type ActorDirection = 'down' | 'left' | 'right' | 'up';

const DIRECTION_CENTER_DEGREES: Record<ActorDirection, number> = {
  down: 90,
  left: 180,
  right: 0,
  up: 270,
};

export function actorDirectionFromHeading(
  headingDegrees: number,
): ActorDirection {
  const normalizedHeading = ((headingDegrees % 360) + 360) % 360;

  if (normalizedHeading >= 315 || normalizedHeading < 45) {
    return 'right';
  }
  if (normalizedHeading < 135) {
    return 'down';
  }
  if (normalizedHeading < 225) {
    return 'left';
  }
  return 'up';
}

export function actorDirectionWithHysteresis(input: {
  currentDirection: ActorDirection;
  headingDegrees: number;
  hysteresisDegrees?: number;
}) {
  const hysteresisDegrees = Math.max(0, input.hysteresisDegrees ?? 10);
  const currentDirectionCenter =
    DIRECTION_CENTER_DEGREES[input.currentDirection];
  const distanceFromCurrentDirection = shortestAngleDistanceDegrees(
    input.headingDegrees,
    currentDirectionCenter,
  );

  return distanceFromCurrentDirection <= 45 + hysteresisDegrees
    ? input.currentDirection
    : actorDirectionFromHeading(input.headingDegrees);
}

function shortestAngleDistanceDegrees(from: number, to: number) {
  return Math.abs(((((to - from) % 360) + 540) % 360) - 180);
}

File: campus-navigate-app/flutter_app/lib/ui/map/models/actor_direction_model.dart
import 'package:indoor_navigation/domain/common/angle_math.dart';
import 'package:indoor_navigation/ui/map/models/actor_models.dart';

const Map<ActorDirection, double> _directionCenterDegrees = {
  ActorDirection.down: 90,
  ActorDirection.left: 180,
  ActorDirection.right: 0,
  ActorDirection.up: 270,
};

ActorDirection actorDirectionFromHeading(double headingDegrees) {
  final normalizedHeading = normalizeDegrees(headingDegrees);
  if (normalizedHeading >= 315 || normalizedHeading < 45) {
    return ActorDirection.right;
  }
  if (normalizedHeading < 135) {
    return ActorDirection.down;
  }
  if (normalizedHeading < 225) {
    return ActorDirection.left;
  }
  return ActorDirection.up;
}

ActorDirection actorDirectionWithHysteresis({
  required ActorDirection currentDirection,
  required double headingDegrees,
  double hysteresisDegrees = 10,
}) {
  final safeHysteresis = hysteresisDegrees < 0 ? 0.0 : hysteresisDegrees;
  final distanceFromCurrentDirection = shortestAngleDistanceDegrees(
    headingDegrees,
    _directionCenterDegrees[currentDirection]!,
  );
  return distanceFromCurrentDirection <= 45 + safeHysteresis
      ? currentDirection
      : actorDirectionFromHeading(headingDegrees);
}
